R/cnd-abort.R

```diff
@@ -248,6 +290,325 @@ signal_abort <- function(cnd, file = NULL) {
+#'
+#' @name args_error_context
+NULL
+
+#' Find the caller argument for error messages
+#'
+#' @description
+#'
+#' `caller_arg()` is a variant of `substitute()` or [ensym()] that
+#' returns an argument name as a string.
+#'
+#' - The resulting name should be formatted as argument, e.g. using
+#'   the `.arg` in the cli package.
+#'
+#' - Use `@inheritParams rlang::args_error_context` to document an
+#'   `arg` or `error_arg` argument that takes `error_arg()` as default.
+#'
+#' @param arg An argument name in the current function.
+#' @usage NULL
+#'
+#' @examples
+#' arg_checker <- function(x, arg = caller_arg(x), call = caller_env()) {
+#'   cli::cli_abort("{.arg arg} must be a thingy.", call = call)
+#' }
+#'
+#' my_function <- function(my_arg) {
+#'   arg_checker(my_arg)
+#' }
+#'
+#' try(my_function(NULL))
+#' @export
+caller_arg <- function(arg) {
+  arg <- substitute(arg)
+  if (!is_symbol(arg)) {
+    abort(sprintf(
+      "%s must be an argument name.",
+      format_arg("arg")
+    ))
+  }
+
+  expr <- do.call(substitute, list(arg), envir = caller_env())
+  if (!is_symbol(expr)) {
+    abort(sprintf(
+      "%s must be an argument name.",
+      format_arg(as_string(arg))
+    ))
+  }
+
+  as_string(expr)
+}
+
+#' Validate and format a function call for use in error messages
+#'
+#' @description
+#'
+#' - `error_call()` creates a function call ready to be used as `call`
+#'   field of error conditions. This field is displayed by [stop()]
+#'   and [abort()] to give context to an error message.
+#'
+#' - `format_error_call()` passes its input to `error_call()` and
+#'   formats the result as code. Use this function if you are
+#'   generating the "in" part of an error message from a stack frame
+#'   call.
+#'
+#' @param call,error_call An expression (as returned by e.g.
+#'   `sys.call()`) representing the context in which the error
+#'   occurred. If non-null, the call is stripped from its arguments to
+#'   keep it simple.
+#'
+#'   Can also be an execution environment of a currently running
+#'   function (as returned by e.g. `parent.frame()`). The
+#'   corresponding call is then retrieved.
+#'
+#'   See also [rlang::local_error_call()].
+#' @return Either a string formatted as code or `NULL` if `call` or
+#'   the result of `error_call(call)` is `NULL`.
+#'
+#' @keywords internal
+#' @export
+error_call <- function(call) {
+  while (is_environment(call)) {
+    flag <- error_flag(call)
+
+    if (is_null(flag) || is_call(flag)) {
+      call <- flag
+      break
+    }
+
+    if (is_environment(flag)) {
+      call <- flag
+      next
+    }
+
+    if (is_string(flag, "caller")) {
+      call <- eval_bare(call2(caller_env), call)
+      next
+    }
+
+    call <- caller_call(call)
+    break
+  }
+
+  # This happens when a function using its caller's call is called in
+  # testthat snapshots or within `local()`.
+  if (is_call(call, c("eval", "evalq"))) {
+    return(NULL)
+  }
+
+  if (is_call(call) && is_expression(call)) {
```

Comment:
What does being a call and an expression mean?

Reply:
The narrow definition of "expression" in rlang is any R object that can be generated by the parser. Being a call and an expression in the narrow sense means that the call does not contain any complex object such as a function definition or inlined vector of length > 1. This check is meant to avoid displaying complex calls created by NSE.

Reply:
I guess to be completely correct we should also check that the CAR is not a call though.

Reply:
The check is now simply:

```r
is_call(call) && is_symbol(call[[1]])
```